### ISSUE TITLE
No I/O Polling on Handle::block_on

### ISSUE THREAD

**Issue by TheBlueMatt** [tokio-rs/tokio#4659]
With a Multi-Threaded executor configured for one thread (I can't seem to reproduce the issue with an MT executor configured to use 2 threads, but that may be a limitation in the application) used in a dual sync-async codebase, I observe the following:

In a sync thread (or a blocking tokio handler thread), a synchronous mutex is taken then Handle::block_on is called. Meanwhile, in the tokio handler thread while processing a network read, the same mutex is locked.

This results in a thread that is parked in Tokio waiting on some IO and a tokio handler thread waiting (indirectly) on the same. However, for some reason, the IO never completes.

This leads me to believe that a block_on call will not poll the IO driver. Is this true, and if not, it seems quite useful for dual sync-async programs for this to be changed.

Also, the docs for Handle::block_on are somewhat mum on this specific issue.

**Comment by Darksonn** [comment]
Only the runtime's own threads will drive the IO driver when using the multi-threaded runtime.

**Comment by Darksonn** [comment]
Are you blocking outside of `block_in_place`? The `block_in_place` method should spawn an additional worker thread to drive the IO driver and other tasks on the runtime. If not, how many parallel calls to `block_in_place` do you have?

**Comment by Darksonn** [comment]
Okay, well. Do not block the runtime thread without using `block_in_place` around the blocking section.

**Comment by Darksonn** [comment]
Even if we make calls to `block_on` able to drive the IO driver, there are lots of other ways your code will break if you block all the runtime threads. In general, if you want to use a blocking synchronous library from async code, it is the async code's responsibility to transition out of the async context using `spawn_blocking` or `block_in_place` every time it calls the sync library.
